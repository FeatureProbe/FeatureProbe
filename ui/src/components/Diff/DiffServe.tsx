import { FormattedMessage } from 'react-intl';
import { CSSProperties } from 'react';
import { Table } from 'semantic-ui-react';
import { ChangeItem } from './diff';
import { DiffFieldValue } from './DiffFieldValue';
import { renderField, renderFieldsItems } from './renderDiff';
import { diffType, positionType } from './constants';
import conditionStyles from './RulesDiffContent.module.scss';
import fieldStyles from './fields.module.scss';
import styles from './DiffServe.module.scss';

interface DiffServeContentProps {
  map?: Map<string, string>;
  content:
    | {
        split?: number[];
        select?: string;
      }
    | ChangeItem[];
  type: positionType;
  diffType: diffType;
  rowStyle?: CSSProperties;
}

export const DiffServeContent: React.FC<DiffServeContentProps> = (props) => {
  const { content, type, diffType } = props;

  const render = () => {
    if (content instanceof Array) {
      return renderField(content, type, (value) => {
        if (value.get('select')) {
          const select = value.get('select');
          if (select?.type === 'add' && type === 'after') {
            return <DiffFieldValue type="add" value={select.value as string} />;
          } else if (select?.type === 'remove' && type === 'before') {
            return <DiffFieldValue type="remove" value={select.value as string} />;
          } else if (select?.type === 'same') {
            return <DiffFieldValue type="same" value={select.value as string} />;
          }
        }
        if (value.get('split')) {
          const split = value.get('split');
          if (split && split.value instanceof Array && typeof split.value[0] !== 'object') {
            return <DiffFieldValue type={split.type} value={split.value} />;
          } else {
            return renderFieldsItems(split?.value as ChangeItem[], type, (value, diffType, _, index) => {
              if (diffType === 'add' && type === 'after') {
                return <DiffFieldValue type="add" value={value as string} key={index} />;
              } else if (diffType === 'remove' && type === 'before') {
                return <DiffFieldValue type="remove" value={value as string} key={index} />;
              } else if (diffType === 'same') {
                return (
                  <DiffFieldValue
                    style={{
                      background: 'rgba(33,37,41,0.08)',
                    }}
                    type="same"
                    value={value as string}
                    key={index}
                  />
                );
              }
            });
          }
        }
      });
    } else {
      if (content.select !== undefined) {
        return <DiffFieldValue value={content.select} />;
      }
      if (content.split !== undefined) {
        return <DiffFieldValue value={content.split as unknown as string[]} />;
      }
    }
  };

  return (
    <Table.Row className={`${fieldStyles[`diff-item-${diffType}`]} ${conditionStyles['condition-diff-item']}`}>
      <Table.Cell>
        <FormattedMessage id="common.serve.text" />
      </Table.Cell>
      <Table.Cell colSpan="5">{render()}</Table.Cell>
    </Table.Row>
  );
};

interface DiffServeProps {
  content:
    | {
        split?: number[];
        select?: string;
      }
    | ChangeItem[];
}

export const DiffServe: React.FC<DiffServeProps> = (props) => {
  const { content } = props;

  return (
    <div className={styles['serve-diff']}>
      <div>
        <Table>
          <Table.Body>
            <DiffServeContent diffType="modify" content={content} type={'before'} />
          </Table.Body>
        </Table>
      </div>
      <div>
        <Table>
          <Table.Body>
            <DiffServeContent diffType="modify" content={content} type={'after'} />
          </Table.Body>
        </Table>
      </div>
    </div>
  );
};
