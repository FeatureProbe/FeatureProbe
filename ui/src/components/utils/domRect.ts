export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}