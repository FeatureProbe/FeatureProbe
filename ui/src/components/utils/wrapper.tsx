import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
