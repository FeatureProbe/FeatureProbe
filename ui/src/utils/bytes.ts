export function getBytes(str: string) {
  let bytes = str.length,
    i = 0;
  for (; i < bytes; i++) if (str.charCodeAt(i) > 255) bytes++;

  return bytes;
}

const MB = 1048576;
export const MAX_SIZE = MB;