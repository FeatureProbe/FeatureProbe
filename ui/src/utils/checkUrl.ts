function escapeStringRegexp(string: string): string {
  if (typeof string !== 'string') {
    throw new TypeError('Expected a string');
  }

  return string
    .replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
    .replace(/-/g, '\\x2d');
}

export function matchUrl(matcher: string, originUrl: string, targetUrl: string): boolean {
  let href = '';
  let hash = '';
  let search = '';

  try {
    const fullUrl = new URL(targetUrl, '');
    href = fullUrl.href;
    hash = fullUrl.hash;
    search = fullUrl.search;
  } catch (error) {
    href = targetUrl;
  }

  let regex;
  let testUrl;

  switch (matcher) {
    case 'EXACT':
      testUrl = href;
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SIMPLE':
      testUrl = href.replace(hash, '').replace(search, '');
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SUBSTRING':
      testUrl = href.replace(search, '');
      regex = new RegExp('.*' + escapeStringRegexp(originUrl) + '.*$');
      break;
    case 'REGULAR':
      testUrl = href.replace(search, '');
      regex = new RegExp(originUrl);
      break;
    default:
      return false;
  }

  return regex.test(testUrl);
}