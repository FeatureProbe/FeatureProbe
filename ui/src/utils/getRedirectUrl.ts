import { getProjectList } from 'services/project';
import { IProject, IEnvironment } from 'interfaces/project';
import { IDictionary } from 'interfaces/targeting';
import { getFromDictionary } from 'services/dictionary';
import { LAST_SEEN } from 'constants/dictionary_keys';

const getInfo = async () => {
  const res =  await getFromDictionary<IDictionary>(LAST_SEEN);
  const { success, data } = res;
  if (success && data) {
    return JSON.parse(data.value);
  } else {
    return {};
  }
};

export const getRedirectUrl = async (defaultRedirectUrl: string) => {
  let redirectUrl = defaultRedirectUrl;
  const info = await getInfo();
  const { projectKey, environmentKey } = info;

  if (!!projectKey && !!environmentKey) {
    redirectUrl = `/${projectKey}/${environmentKey}/toggles`;
  } else {
    const res = await getProjectList<IProject[]>();
    const { success, data } = res;
    if (success && data) {
      if (data[0]) {
        const { key, environments } = data[0];
        let envKey = '';
        environments.some((environment: IEnvironment) => {
          if (environment.key) {
            envKey = environment.key;
            return true;
          }
          return environment;
        });
  
        if (key && envKey) {
          redirectUrl = `/${key}/${envKey}/toggles`;
        }
      } else {
        redirectUrl = '/projects';
      }
    }
  }

  return redirectUrl;
};