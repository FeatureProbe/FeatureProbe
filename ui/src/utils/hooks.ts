
import { useState } from 'react';

export const useLocalStorage = (key: string, initialValue: boolean | string) => {
  const [storedValue, setStoredValue] = useState(() => {
    if (typeof window === 'undefined') {
      return initialValue;
    }

    try {
      const item = window.localStorage.getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      console.log(error);
      return initialValue;
    }
  });

  const setValue = (value: object) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);

      if (typeof window !== 'undefined') {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
      
    } catch (error) {
      console.log(error);
    }
  };
  return [storedValue, setValue];
};
