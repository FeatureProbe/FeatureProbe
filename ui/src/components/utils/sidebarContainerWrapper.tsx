import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
