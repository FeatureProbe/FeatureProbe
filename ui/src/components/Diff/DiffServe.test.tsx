import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { DiffServe } from './DiffServe';
import idiff from './diff';
import { IntlWrapper } from 'components/utils/wrapper';

const content_a = idiff(
  {
    select: 'name 1',
  },
  {
    select: 'name 2',
  }
);

const content_b = idiff(
  {
    select: 'name 1',
  },
  {
    split: [50, 20],
  }
);

const content_c = idiff(
  {
    select: 'name 1',
  },
  {
    select: 'name 1',
  }
);

const content_d = idiff(
  {
    split: [50, 50],
  },
  {
    split: [40, 60],
  }
);

const content_e = {
  split: [10, 90]
};

const content_f = {
  select: 'name 1'
};

const content_g = idiff(
  {
    split: [50, 50],
  },
  {
    split: [50, 60],
  }
);

it('DiffServe snapshot', (done) => {
  (async () => {
    const { asFragment, rerender } = render(<DiffServe content={content_a} />, {
      wrapper: IntlWrapper,
    });
    expect(asFragment()).toMatchSnapshot();

    rerender(<DiffServe content={content_b} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(<DiffServe content={content_c} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(<DiffServe content={content_d} />);
    expect(asFragment()).toMatchSnapshot();
    
    rerender(<DiffServe content={content_e} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(<DiffServe content={content_f} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(<DiffServe content={content_g} />);
    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});
