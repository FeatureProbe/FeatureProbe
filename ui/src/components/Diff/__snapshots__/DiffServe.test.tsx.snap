// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffServe snapshot 1`] = `
<DocumentFragment>
  <div
    class="serve-diff"
  >
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-remove"
              >
                name 1
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-add"
              >
                name 2
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffServe snapshot 2`] = `
<DocumentFragment>
  <div
    class="serve-diff"
  >
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-remove"
              >
                name 1
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div>
                <div
                  class="diff-field-value field-value-add"
                >
                  50
                </div>
                <div
                  class="diff-field-value field-value-add"
                >
                  20
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffServe snapshot 3`] = `
<DocumentFragment>
  <div
    class="serve-diff"
  >
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-same"
              >
                name 1
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-same"
              >
                name 1
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffServe snapshot 4`] = `
<DocumentFragment>
  <div
    class="serve-diff"
  >
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-remove"
              >
                50
              </div>
              <div
                class="diff-field-value field-value-remove"
              >
                50
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-add"
              >
                40
              </div>
              <div
                class="diff-field-value field-value-add"
              >
                60
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffServe snapshot 5`] = `
<DocumentFragment>
  <div
    class="serve-diff"
  >
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div>
                <div
                  class="diff-field-value field-value-undefined"
                >
                  10
                </div>
                <div
                  class="diff-field-value field-value-undefined"
                >
                  90
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div>
                <div
                  class="diff-field-value field-value-undefined"
                >
                  10
                </div>
                <div
                  class="diff-field-value field-value-undefined"
                >
                  90
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffServe snapshot 6`] = `
<DocumentFragment>
  <div
    class="serve-diff"
  >
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-undefined"
              >
                name 1
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-undefined"
              >
                name 1
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DiffServe snapshot 7`] = `
<DocumentFragment>
  <div
    class="serve-diff"
  >
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-same"
                style="background: rgba(33, 37, 41, 0.08);"
              >
                50
              </div>
              <div
                class="diff-field-value field-value-remove"
              >
                50
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <table
        class="ui table"
      >
        <tbody
          class=""
        >
          <tr
            class="diff-item-modify condition-diff-item"
          >
            <td
              class=""
            >
              Serve
            </td>
            <td
              class=""
              colspan="5"
            >
              <div
                class="diff-field-value field-value-same"
                style="background: rgba(33, 37, 41, 0.08);"
              >
                50
              </div>
              <div
                class="diff-field-value field-value-add"
              >
                60
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
